### PR TITLE
refactor: allow multipart default commands

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
In case a command like `sample|command|*default` is registered CLI should be able to correctly execute 
`tns sample command`

Added tests to ensure functionality

Ping @rosen-vladimirov @Fatme

### CODE REVIEW

File: yok.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import * as path from "path";
 import { annotate, isPromise } from "./helpers";
 import { ERROR_NO_VALID_SUBCOMMAND_FORMAT } from "./constants";
+const CommandSeparator = "|";
```
Comment (rosen-vladimirov) [inline yok.ts]: You can use the `CommandsDelimiters` from constants

File: yok.ts [hunk 2]
```diff
@@ -160,9 +162,24 @@ export class Yok implements IInjector {
...
-		const defaultCommand = _.find(subCommands, (command) => _.startsWith(command, "*"));
+		const defaultCommand = _.find(subCommands, command => _.some(command.split(CommandSeparator), c => _.startsWith(c, DefaultCommandSymbol)));
+
+		if (defaultCommand) {
+			// In case the defaultCommand consists of more than one part (e.g. someCommand|someSubcommand|*default)
+			// we need to remove all parts from the arguments list - they are not actually arguments
+			defaultCommand
+				.split(CommandSeparator)
+				.map(c => _.startsWith(c, DefaultCommandSymbol) ? c.substr(1) : c)
+				.forEach(a => {
```
Comment (rosen-vladimirov) [inline yok.ts]: can you use better names.. a and c are not meaningful

File: yok.ts [hunk 3]
```diff
@@ -176,7 +193,7 @@ export class Yok implements IInjector {
 			arg = arg.toLowerCase();
 			currentSubCommandName = currentSubCommandName ? this.getHierarchicalCommandName(currentSubCommandName, arg) : arg;
 			remainingArguments = _.tail(remainingArguments);
-			if (matchingSubCommandName = _.find(subCommands, (sc) => sc === currentSubCommandName || sc === "*" + currentSubCommandName)) {
+			if (matchingSubCommandName = _.find(subCommands, (sc) => sc === currentSubCommandName || sc === DefaultCommandSymbol + currentSubCommandName)) {
```
Comment (rosen-vladimirov) [inline yok.ts]: You can use string interpolation here

File: test/unit-tests/yok.ts [hunk 4]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+		let injector: IInjector;
+		beforeEach(() => {
+			injector = new Yok();
+		});
+
+		describe("returns true", () => {
+			describe("when command consists of two parts", () => {
+				it("and is valid", async () => {
+					injector.requireCommand("sample|command", "sampleFileName");
+					assert.isTrue(await injector.isValidHierarchicalCommand("sample", ["command"]));
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: You can use:
```TypeScript
await assert.eventually.isTrue(injector.isValidHierarchicalCommand("sample", ["command"]));
```

But this is not a merge stopper

File: test/unit-tests/yok.ts [hunk 5]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+				});
+
+				it("and has default value and no arguments passed", async () => {
+					injector.requireCommand("sample|*default", "sampleFileName");
+					assert.isTrue(await injector.isValidHierarchicalCommand("sample", []));
+				});
+
+				it("and has default value and canExecute returns true", async () => {
+					const command = {
+						canExecute: () => true
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: canExecute must return Promise.

File: test/unit-tests/yok.ts [hunk 6]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+					};
+					injector.registerCommand("sample|*default", command);
+					injector.requireCommand("sample|*default", "sampleFileName");
+					assert.isTrue(await injector.isValidHierarchicalCommand("sample", ["arg"]));
+				});
+
+				it("and has default value and allowedParameters' length is greater than 0", async () => {
+					const command = {
+						allowedParameters: ["param1", "param2"]
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: allowedParameters is array of ICommandParameter, not strings

File: test/unit-tests/yok.ts [hunk 7]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+
+		describe("returns false", () => {
+			it("when command is invalid", async () => {
+				injector.requireCommand("sample|command", "sampleFileName");
+				assert.isFalse(await injector.isValidHierarchicalCommand("wrong", ["command"]));
+			});
+
+			it("when has default value and canExecute returns false", async () => {
+				const command = {
+					canExecute: () => false
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: this should return Promise

File: test/unit-tests/yok.ts [hunk 8]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+			});
+
+			it("when has default value and allowedParameters' length is 0", async () => {
+				const allowedParameters: string[] = [];
+				const command = {
+					allowedParameters
+				};
+				injector.registerCommand("sample|*default", command);
+				injector.requireCommand("sample|*default", "sampleFileName");
+				assert.isRejected(injector.isValidHierarchicalCommand("sample", ["arg"]));
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: missing await or return here

File: test/unit-tests/yok.ts [hunk 9]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+
+		describe("throws", () => {
+			it("when has default value and some arguments passed", async () => {
+				injector.requireCommand("sample|*default", "sampleFileName");
+				assert.isRejected(injector.isValidHierarchicalCommand("sample", ["argument"]));
+			});
+
+			it("when arguments are invalid", async () => {
+				injector.requireCommand("sample|command", "sampleFileName");
+				assert.isRejected(injector.isValidHierarchicalCommand("sample", ["commandarg"]));
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: missing await or return here

File: test/unit-tests/yok.ts [hunk 10]
```diff
@@ -515,6 +515,105 @@ $injector.register("a", A);
...
+				injector.registerCommand("sample|*default", command);
+				injector.requireCommand("sample|*default", "sampleFileName");
+				assert.isFalse(await injector.isValidHierarchicalCommand("sample", ["arg"]));
+			});
+		});
+
+		describe("throws", () => {
+			it("when has default value and some arguments passed", async () => {
+				injector.requireCommand("sample|*default", "sampleFileName");
+				assert.isRejected(injector.isValidHierarchicalCommand("sample", ["argument"]));
```
Comment (rosen-vladimirov) [inline test/unit-tests/yok.ts]: missing await or return here

File: yok.ts (hunk 2 above)
Comment (Fatme) [inline yok.ts]: Can we use
```
.map(c => _.startsWith(c, DefaultCommandSymbol) ? c.substr(1) : c)
.forEach(command => _.remove(commandArguments, command))
```

File: yok.ts [hunk 11]
```diff
@@ -210,7 +227,7 @@ export class Yok implements IInjector {
 							commandName = defaultCommand ? this.getHierarchicalCommandName(name, defaultCommand) : "help";
 							// If we'll execute the default command, but it's full name had been written by the user
 							// for example "appbuilder cloud list", we have to remove the "list" option from the arguments that we'll pass to the command.
-							if (_.includes(this.hierarchicalCommands[name], "*" + args[0])) {
+							if (_.includes(this.hierarchicalCommands[name], DefaultCommandSymbol + args[0])) {
```
Comment (Fatme) [inline yok.ts]:
```
`${DefaultCommandSymbol}${args[0]}`
```

**Pull request merged**